### PR TITLE
HPCC-31159 ECL Watch v9 fix file rename issue

### PR DESCRIPTION
**Pull request by jeclrsg**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/src/src-react/components/forms/RenameFile.tsx [hunk 1]
```diff
@@ -47,21 +54,53 @@ export const RenameFile: React.FunctionComponent<RenameFileProps> = ({
...
+                        const request = { ...data, srcname: logicalFiles[0], DFUServerQueue: "" };
+                        service.Rename(request).then(async response => {
+                            const wuid = response?.wuid ?? null;
+                            if (!wuid) {
+                                logger.error(response?.Exceptions?.Exception[0]?.Message);
+                                return;
+                            }
+                            const dfuWUResponse = await service.GetDFUWorkunit({ wuid });
+                            const { ID, StateMessage = "", SummaryMessage = "" } = { ...dfuWUResponse?.result };
+                            if (StateMessage === FilesprayStates[5]) { // failed
```
Comment (GordonSmith) [inline esp/src/src-react/components/forms/RenameFile.tsx:66]: Should really add a States enum into @hpcc-js/comms?

File: esp/src/src-react/components/forms/RenameFile.tsx [hunk 2]
```diff
@@ -47,21 +54,53 @@ export const RenameFile: React.FunctionComponent<RenameFileProps> = ({
...
+                            Promise.all(dfuRequests).then(responses => {
+                                responses.forEach(response => {
+                                    const StateMessage = response?.result?.StateMessage ?? FilesprayStates[999];
+                                    const ID = response?.result?.ID;
+                                    if (StateMessage === FilesprayStates[5]) { // failed
+                                        logger.error(response?.result?.SummaryMessage ?? "");
+                                    } else if (ID) {
+                                        window.setTimeout(function () { window.open(`#/dfuworkunits/${ID}`); }, 0);
+                                    }
+                                });
```
Comment (GordonSmith) [inline esp/src/src-react/components/forms/RenameFile.tsx:103]: This code would work equaly well for the case of `logicalFiles?.length===1` - except the bit at the end where you open a new tab or "navigate to"?

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: esp/src/src-react/components/forms/RenameFile.tsx [hunk 3]
```diff
@@ -38,42 +42,71 @@ export const RenameFile: React.FunctionComponent<RenameFileProps> = ({
...
-                                closeForm();
-                                if (refreshGrid) refreshGrid(true);
-                            });
-                        });
+            async (data, evt) => {
+                const renameRequests = [];
+                const getDfuWuRequests = [];
+
+                if (logicalFiles?.length === 1) {
+                    const request = { ...data, srcname: logicalFiles[0], DFUServerQueue: "" };
```
Comment (GordonSmith) [inline esp/src/src-react/components/forms/RenameFile.tsx:58]: Is `data.targetRenameFile` an array or a string here?

File: esp/src/src-react/components/forms/RenameFile.tsx [hunk 4]
```diff
@@ -38,42 +42,71 @@ export const RenameFile: React.FunctionComponent<RenameFileProps> = ({
...
+                    renameRequests.push(service.Rename(request));
+                } else {
+                    logicalFiles.forEach((logicalFile, idx) => {
+                        const request = { ...data, srcname: logicalFile, dstname: data.targetRenameFile[idx].name, DFUServerQueue: "" };
```
Comment (GordonSmith) [inline esp/src/src-react/components/forms/RenameFile.tsx:62]: ...`data.targetRenameFile` is an array here...

File: esp/src/src-react/components/forms/RenameFile.tsx [hunk 5]
```diff
@@ -38,42 +42,71 @@ export const RenameFile: React.FunctionComponent<RenameFileProps> = ({
...
+                } else {
+                    getDfuWuResponses.forEach(response => {
+                        const State = response?.result?.State ?? FileSprayStates.unknown;
+                        const ID = response?.result?.ID;
+                        if (State === FileSprayStates.failed) {
+                            logger.error(response?.result?.SummaryMessage ?? "");
+                        } else if (ID) {
+                            window.setTimeout(function () { window.open(`#/dfuworkunits/${ID}`); }, 0);
+                        }
+                    });
```
Comment (GordonSmith) [inline esp/src/src-react/components/forms/RenameFile.tsx:100]: This if / else has different logic for error handling?  Could be combined to a single flow with the exception of the `window.open` and `pushUrl`?

**Review by GordonSmith — CHANGES_REQUESTED** [review]
The rename logic and error handling looks different for the case when there is 1 v many renames?

File: esp/src/src-react/components/forms/RenameFile.tsx (hunk 3 above)
Comment (jeclrsg) [reply esp/src/src-react/components/forms/RenameFile.tsx]: It was undefined. I don't recall why, but if there was one file the input was named dstname, and if there was a list, it used the targetRenameFile array. It only uses the targetRenameFile array now.

**Review by GordonSmith — APPROVED** [review]

**Pull request merged**